### ISSUE TITLE
File/directory names with a question mark character in them not handled

### ISSUE THREAD

**Issue by pepa65** [ltworf/weborf#7]
I really hope this is fixable. When running weborf listing files & directories, it cannot process any names that have a question mark in them. Directories cannot be descended into, files cannot be viewed/downloaded. It says: Error 404, page not found.

**Comment by ltworf** [comment]
https://en.wikipedia.org/wiki/Query_string

I think it's the desired behaviour. `?` is supposed to separate between resource name and parameters.

**Comment by ltworf** [comment]
I think possibly, it would have to handle the escaped `?` separately from the real `?`, but I need to check what the specification is saying.

**Comment by pepa65** [comment]
I understand the role of ? in URLs. But weborf can know when it is actually part of a filename that it has read, and in that case, it would be great if these kind of filenames can be used as well.

I don't know how hard it would be to record that information in the current design, whether it is a filename or a URL, because in the end, it all becomes a URL...

**Comment by pepa65** [comment]
Super, thank you very much. That wasn't very easy...
I found I had to define ESCAPED_FNAME_LEN as well.

**Comment by ltworf** [comment]
There is still the issue that filenames with `%` don't work. I am almost sure that it is a bug in the function that replaces escapes with the real character. Maybe you could have a look at that?

**Comment by pepa65** [comment]
Well done. I actually haven't managed to backport this successfully... It looks like it should work the same. It will come.

As to your present version, apart from the %, there is one other issue which should be easy to solve: the names that are rendered in html might get mangled by the html processing. I had one name `a&quot;b` which was rendered `a'b` and a name `a<b` which was rendered `a` -- both worked fine when you click on it.

About that function with the long list of problematic characters, I've found that only these gave problems:
# and % (404) and : (browser confused about the url). Maybe you can leave all other ones out? I have too little understanding to solve the % problem for now...

**Comment by pepa65** [comment]
I found the problem with the `%`: in `replaceEscape` (`mystring.c`) when the %hh is found and replaced, string is still pointing to it. When it was replaced by a `%`, it will process it again... So `string++` after the replacement should fix that.

**Comment by ltworf** [comment]
done

**Comment by pepa65** [comment]
Not sure if I should open a new Issue, but there are some remaining problems:
* Rendering of filenames with `<` in them cuts all characters starting from < (but opening them works fine).
* The href to filenames with `"` in them is cut similarly, and of course they can't be opened like that, but the links work fine when manually typed in (the rendering is fine).
* For testing, I made a directory name ` !"#$%&'()*+,:;<=>?@\|[]/`. The href is not properly encoded, because the `<` and `>` are not %hh encoded.
But it seems we're almost there!

**Comment by pepa65** [comment]
This is a format-patch with the proposed solution:
```
From 2dcc7498ca95b37b35909cba9278a0ca0e300b7e Mon Sep 17 00:00:00 2001
From: pepa65 <solusos@passchier.net>
Date: Sat, 25 Feb 2017 00:44:03 +0700
Subject: [PATCH] Added html encoding for rendering of problematic filenames
 Added characters to be %hh encoded for hrefs

---
 utils.c | 51 +++++++++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 49 insertions(+), 2 deletions(-)

diff --git a/utils.c b/utils.c
index d666ec0..d3e0f0b 100644
--- a/utils.c
+++ b/utils.c
@@ -75,6 +75,10 @@ static void uri_encode(char *dest, size_t size, char *origin) {
             origin[i] == '[' ||
             origin[i] == ']' ||
             origin[i] == '%' ||
+            origin[i] == '>' ||
+            origin[i] == '<' ||
+            origin[i] == '\"' ||
+            origin[i] == ' ' ||
             origin[i] == '\\'
         ) {
             format = "%%%02x";
@@ -88,6 +92,46 @@ static void uri_encode(char *dest, size_t size, char *origin) {
 }
 
 /**
+ * This function encodes the characters in 'origin' to &code; form into 'encoded'
+ **/
+static void html_encode(char *dest, size_t size, char *origin) {
+    if (size == 0 || dest == NULL || origin == NULL) return;
+
+    dest[0] = '\0';
+
+    size_t len = 0;
+    size_t inc;
+    char *code;
+    size_t codesize;
+    char ni[2] = "X";
+
+    for (size_t i = 0; i < strlen(origin); i++) {
+        ni[0] = origin[i];
+        code = ni;
+        inc = 1;
+        switch (origin[i]) {
+        case '&':
+            code = "&amp;";
+            break;
+        case '<':
+            code = "&lt;";
+            break;
+        case '>':
+            code = "&gt;";
+            break;
+        case ' ':
+            code = "&nbsp;";
+            break;
+        }
+        codesize = strlen(code);
+        if (len + codesize >= ESCAPED_FNAME_LEN)
+            return;
+        strcpy(dest + len, code);
+        len += codesize;
+    }
+}
+
+/**
 This function reads the directory dir, putting inside the html string an html
 page with links to all the files within the directory.
 
@@ -114,6 +158,7 @@ int list_dir(connection_t *connection_prop, char *html, unsigned int bufsize, bo
 
     struct dirent **namelist = NULL;
     counter = scandir(connection_prop->strfile, &namelist, 0, alphasort);
+    char *name_html = malloc(ESCAPED_FNAME_LEN);
     char *escaped_dname = malloc(ESCAPED_FNAME_LEN);
 
 
@@ -156,6 +201,7 @@ int list_dir(connection_t *connection_prop, char *html, unsigned int bufsize, bo
         localtime_r(&f_prop.st_mtime,&ts);
         strftime(last_modified,URI_LEN, "%a, %d %b %Y %H:%M:%S", &ts);
 
+        html_encode(name_html, ESCAPED_FNAME_LEN, namelist[i]->d_name);
         uri_encode(escaped_dname, ESCAPED_FNAME_LEN, namelist[i]->d_name);
 
         if (S_ISREG(f_mode)) { //Regular file
@@ -181,7 +227,7 @@ int list_dir(connection_t *connection_prop, char *html, unsigned int bufsize, bo
                 color = "#EAEAEA";
             printf_s=snprintf(html+pagesize,maxsize,
                               "<tr style=\"background-color: %s;\"><td>f</td><td><a href=\"%s\">%s</a></td><td>%lld%s</td><td>%s</td></tr>\n",
-                              color, escaped_dname, namelist[i]->d_name, (long long int)size, measure,last_modified);
+                              color, escaped_dname, name_html, (long long int)size, measure,last_modified);
             maxsize-=printf_s;
             pagesize+=printf_s;
 
@@ -189,7 +235,7 @@ int list_dir(connection_t *connection_prop, char *html, unsigned int bufsize, bo
             //Table row for the dir
             printf_s=snprintf(html+pagesize,maxsize,
                               "<tr style=\"background-color: #DFDFDF;\"><td>d</td><td><a href=\"%s/\">%s/</a></td><td>-</td><td>%s</td></tr>\n",
-                              escaped_dname, namelist[i]->d_name,last_modified);
+                              escaped_dname, name_html,last_modified);
             maxsize-=printf_s;
             pagesize+=printf_s;
         }
@@ -201,6 +247,7 @@ int list_dir(connection_t *connection_prop, char *html, unsigned int bufsize, bo
     }
 
 escape:
+    free(name_html);
     free(escaped_dname);
     free(namelist);
     if (errcode == 0) {
-- 
2.7.4
```

**Comment by ltworf** [comment]
This should go in a separate issue, but thanks for the patch, I'll review and merge it when I have some time.

**Comment by ltworf** [comment]
I merged the patch, however there are some remarks.

There is an unused variable, and there is a strlen() in a loop, this makes it much more slow.

I made another commit to change the char* to a const char * so that the compiler can optimise and run the strlen only once, but as a general rule you don't want to put strlen in a loop condition, or it will parse the entire string on every iteration of the loop.

**Comment by pepa65** [comment]
Thanks for the feedback. Indeed, at first I didn't have `size`, and used the precompiler constant, but I realized that is more fragile, less encapsulated. I forgot to actually use/change it... Good for me to understand about const. I guess the size parameters could also be const?

**Comment by ltworf** [comment]
size is an int, i think the compiler can figure it out, the issue is more for pointers, telling the compiler that the function doesn't change the pointed area

**Comment by pepa65** [comment]
Theoretically, a fname that has been converted to 'html' could be up to 6 times the size of the fname... (`&nbsp;` times 256), so I changed the allocation for the html names on principle.
